File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/chain_of_responsibility/solution/CustomerServiceSupervisor.m
% CustomerServiceSupervisor.m

classdef CustomerServiceSupervisor < Handler
	methods
		% constructor:
		function theCustomerServiceSupervisor = CustomerServiceSupervisor(aHandler)
			theCustomerServiceSupervisor@Handler(aHandler);
		end
	end

	methods (Access = protected)
		function result = canHandle(theCustomerServiceSupervisor, aCustomerRequestsDiscountEvent)
			result = (aCustomerRequestsDiscountEvent.Percent < 20);
		end

		function doHandle(theCustomerServiceSupervisor, aCustomerRequestsDiscountEvent)
			fprintf('I am a CustomerServiceSupervisor.\n');
			fprintf('I can give you a %d%% discount.\n', aCustomerRequestsDiscountEvent.Percent);
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/chain_of_responsibility/solution/CustomerServiceSupervisor.java
// CustomerServiceSupervisor.java

public class CustomerServiceSupervisor extends Handler
{
	public CustomerServiceSupervisor(Handler aSuccessor)
	{
		super(aSuccessor);
	}

	protected boolean canHandle(CustomerRequestsDiscountEvent aCustomerRequestsDiscountEvent)
	{
		return (aCustomerRequestsDiscountEvent.getPercent() < 20);
	}

	protected void doHandle(CustomerRequestsDiscountEvent aCustomerRequestsDiscountEvent)
	{
		System.out.println("I am a CustomerServiceSupervisor.");
		System.out.println("I can give you a " + aCustomerRequestsDiscountEvent.getPercent() + "% discount.");
	}
}
